ua-snap/data-api
# Explicit argument validation

This PR addresses closes #337 by using endpoint-specific if/then routines to validate arguments. Below is a comprehensive list of requests to test for each affected endpoint. Feel free to pick and choose / create new requests of your own for testing. (Side note, I originally thought the hydrology endpoint would need this treatment, but it looks like it was already done in previous work.)

Is this too much extra code? I have a feeling that the conditional statements could be written more succinctly / efficiently, 
so if you see any obvious areas for improvement / code-trimming please advise!

**TO TEST:** 

Click the links below and confirm expected behavior. Or, create your own requests. It might also be helpful to have a browser open with the production version of the API to compare the returns. In my experience, all the requests below in the "400 club" currently return data in the production version. 

# **degree days:**
these should work:
http://127.0.0.1:5000/degree_days/heating/65.0628/-146.1627?summarize=mmm
http://127.0.0.1:5000/degree_days/heating/65.0628/-146.1627?format=csv
http://127.0.0.1:5000/degree_days/heating/65.0628/-146.1627?summarize=mmm&format=csv
http://127.0.0.1:5000/degree_days/heating/65.0628/-146.1627/1990/2080?summarize=mmm
http://127.0.0.1:5000/degree_days/heating/65.0628/-146.1627/1990/2080?format=csv
http://127.0.0.1:5000/degree_days/heating/65.0628/-146.1627/1990/2080?summarize=mmm&format=csv

these should return 400:
http://127.0.0.1:5000/degree_days/heating/65.0628/-146.1627?summarize=mm
http://127.0.0.1:5000/degree_days/heating/65.0628/-146.1627?summary=mmm 
http://127.0.0.1:5000/degree_days/heating/65.0628/-146.1627?forma=csv
http://127.0.0.1:5000/degree_days/heating/65.0628/-146.1627?format=cs
http://127.0.0.1:5000/degree_days/heating/65.0628/-146.1627?summarize=mm&format=csv
http://127.0.0.1:5000/degree_days/heating/65.0628/-146.1627?summarize=mmm&format=cs
http://127.0.0.1:5000/degree_days/heating/65.0628/-146.1627/1990/2080?summarize=mm
http://127.0.0.1:5000/degree_days/heating/65.0628/-146.1627/1990/2080?summary=mmm 
http://127.0.0.1:5000/degree_days/heating/65.0628/-146.1627/1990/2080?forma=csv
http://127.0.0.1:5000/degree_days/heating/65.0628/-146.1627/1990/2080?format=cs
http://127.0.0.1:5000/degree_days/heating/65.0628/-146.1627/1990/2080?summarize=mm&format=csv
http://127.0.0.1:5000/degree_days/heating/65.0628/-146.1627/1990/2080?summarize=mmm&format=cs

# **snow**:
these should work:
http://127.0.0.1:5000/snow/snowfallequivalent/61.5/-147?summarize=mmm
http://127.0.0.1:5000/snow/snowfallequivalent/61.5/-147?format=csv

these should return 400:
http://127.0.0.1:5000/snow/snowfallequivalent/61.5/-147?summarize=mm
http://127.0.0.1:5000/snow/snowfallequivalent/61.5/-147?summary=mmm
http://127.0.0.1:5000/snow/snowfallequivalent/61.5/-147?forma=csv
http://127.0.0.1:5000/snow/snowfallequivalent/61.5/-147?format=cs

# **taspr**:
these should work:
http://127.0.0.1:5000/temperature/point/65.0628/-146.1627?format=csv
http://127.0.0.1:5000/precipitation/point/65.0628/-146.1627?format=csv
http://127.0.0.1:5000/taspr/point/65.0628/-146.1627?format=csv
http://127.0.0.1:5000/temperature/area/19010208?format=csv
http://127.0.0.1:5000/precipitation/area/19010208?format=csv
http://127.0.0.1:5000/taspr/area/19010208?format=csv
http://127.0.0.1:5000/precipitation/frequency/point/65.028/-146.1627?format=csv
http://127.0.0.1:5000/temperature/65.0628/-146.1627?format=csv
http://127.0.0.1:5000/temperature/65.0628/-146.1627?summarize=mmm
http://127.0.0.1:5000/precipitation/65.0628/-146.1627?format=csv
http://127.0.0.1:5000/precipitation/65.0628/-146.1627?summarize=mmm
http://127.0.0.1:5000/temperature/jan/65.0628/-146.1627?format=csv
http://127.0.0.1:5000/temperature/jan/65.0628/-146.1627?summarize=mmm
http://127.0.0.1:5000/temperature/65.0628/-146.1627/1940/2060?format=csv
http://127.0.0.1:5000/temperature/65.0628/-146.1627/1940/2060?summarize=mmm
http://127.0.0.1:5000/precipitation/65.0628/-146.1627/1940/2060?format=csv
http://127.0.0.1:5000/precipitation/65.0628/-146.1627/1940/2060?summarize=mmm
http://127.0.0.1:5000/temperature/jan/65.0628/-146.1627/1940/2060?format=csv
http://127.0.0.1:5000/temperature/jan/65.0628/-146.1627/1940/2060?summarize=mmm


these should return 400:
http://127.0.0.1:5000/temperature/point/65.0628/-146.1627?forma=csv
http://127.0.0.1:5000/temperature/point/65.0628/-146.1627?format=cs
http://127.0.0.1:5000/precipitation/point/65.0628/-146.1627?forma=csv
http://127.0.0.1:5000/precipitation/point/65.0628/-146.1627?format=cs
http://127.0.0.1:5000/taspr/point/65.0628/-146.1627?forma=csv
http://127.0.0.1:5000/taspr/point/65.0628/-146.1627?format=cs
http://127.0.0.1:5000/temperature/area/19010208?forma=csv
http://127.0.0.1:5000/temperature/area/19010208?format=cs
http://127.0.0.1:5000/precipitation/area/19010208?forma=csv
http://127.0.0.1:5000/precipitation/area/19010208?format=cs
http://127.0.0.1:5000/taspr/area/19010208?forma=csv
http://127.0.0.1:5000/taspr/area/19010208?format=cs
http://127.0.0.1:5000/precipitation/frequency/point/65.028/-146.1627?forma=csv
http://127.0.0.1:5000/precipitation/frequency/point/65.028/-146.1627?format=cs
http://127.0.0.1:5000/temperature/65.0628/-146.1627?forma=csv
http://127.0.0.1:5000/temperature/65.0628/-146.1627?format=cs
http://127.0.0.1:5000/temperature/65.0628/-146.1627?summary=mmm
http://127.0.0.1:5000/temperature/65.0628/-146.1627?summarize=mm
http://127.0.0.1:5000/precipitation/65.0628/-146.1627?forma=csv
http://127.0.0.1:5000/precipitation/65.0628/-146.1627?format=cs
http://127.0.0.1:5000/temperature/65.0628/-146.1627?summary=mmm
http://127.0.0.1:5000/temperature/65.0628/-146.1627?summarize=mm
http://127.0.0.1:5000/temperature/jan/65.0628/-146.1627?forma=csv
http://127.0.0.1:5000/temperature/jan/65.0628/-146.1627?format=cs
http://127.0.0.1:5000/temperature/65.0628/-146.1627?summary=mmm
http://127.0.0.1:5000/temperature/65.0628/-146.1627?summarize=mm
http://127.0.0.1:5000/temperature/65.0628/-146.1627/1940/2060?forma=csv
http://127.0.0.1:5000/temperature/65.0628/-146.1627/1940/2060?format=cs
http://127.0.0.1:5000/temperature/65.0628/-146.1627/1940/2060?summary=mmm
http://127.0.0.1:5000/temperature/65.0628/-146.1627/1940/2060?summarize=mm
http://127.0.0.1:5000/precipitation/65.0628/-146.1627/1940/2060?forma=csv
http://127.0.0.1:5000/precipitation/65.0628/-146.1627/1940/2060?format=cs
http://127.0.0.1:5000/temperature/65.0628/-146.1627/1940/2060?summary=mmm
http://127.0.0.1:5000/temperature/65.0628/-146.1627/1940/2060?summarize=mm
http://127.0.0.1:5000/temperature/jan/65.0628/-146.1627/1940/2060?forma=csv
http://127.0.0.1:5000/temperature/jan/65.0628/-146.1627/1940/2060?format=cs
http://127.0.0.1:5000/temperature/65.0628/-146.1627/1940/2060?summary=mmm
http://127.0.0.1:5000/temperature/65.0628/-146.1627/1940/2060?summarize=mm
